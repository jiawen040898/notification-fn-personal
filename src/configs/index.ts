export * from './aws.config';
export * from './pusher.config';
