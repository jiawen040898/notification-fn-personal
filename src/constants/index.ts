export * from './app-activity-type.enum';
export * from './assignment-due-variable-type.enum';
export * from './cache-store.enum';
export * from './email-template.enum';
export * from './event-type.enum';
export * from './icon-type.enum';
export * from './locale.enum';
