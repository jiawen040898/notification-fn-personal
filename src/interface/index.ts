export * from './company.interface';
export * from './email-request';
export * from './event-model.interface';
export * from './feedback.interface';
export * from './identity.interface';
export * from './manager-subordinate-aggregated-report-data.interface';
export * from './program.interface';
export * from './push-notification-request.interface';
export * from './ses-notification';
