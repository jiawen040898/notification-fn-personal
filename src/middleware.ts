import middy from '@middy/core';
import { initSentryMiddleware, loggerMiddleware } from '@pulsifi/fn';
import * as Sentry from '@sentry/serverless';
import { Handler } from 'aws-lambda';

import { version } from '../package.json';

initSentryMiddleware(version);

export const eventMiddleware = (handler: Handler): Handler =>
    Sentry.AWSLambda.wrapHandler(
        middy(handler).use([
            // ssm({
            //     setToContext: true,
            //     fetchData: {
            //         pusherBeamsInstanceId:
            //             '/notification-fn/PUSHER_BEAMS_INSTANCE_ID',
            //         pusherBeamsSecretKey:
            //             '/notification-fn/PUSHER_BEAMS_SECRET_KEY',
            //     },
            // }),
            loggerMiddleware(),
        ]),
        {
            timeoutWarningLimit: 5.5,
        },
        // .before(async (request) => {
        //     const data = await getInternal(
        //         ['pusherBeamsInstanceId', 'pusherBeamsSecretKey'],
        //         request,
        //     );
        //
        //     PusherConfig.beamsInstanceId = data.pusherBeamsInstanceId;
        //     PusherConfig.beamsSecretKey = data.pusherBeamsSecretKey;
        // }),
    );
