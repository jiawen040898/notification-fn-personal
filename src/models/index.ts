export * from './app-activity-message.entity';
export * from './app-message-template.entity';
export * from './email-activity-message.entity';
export * from './email-template.entity';
export * from './entities';
export * from './user-notification-setting.entity';
