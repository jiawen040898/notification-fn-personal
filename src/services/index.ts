export * from './app-notification.service';
export * from './company.service';
export * from './company-email-template.service';
export * from './email-notification.service';
export * from './email-tracking.service';
export * from './identity.service';
export * from './push-notification.service';
export * from './send-email.service';
export * from './talent-management-notification.service';
