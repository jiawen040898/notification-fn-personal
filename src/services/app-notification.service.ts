import { EventModel } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ApplicationWithdraw } from '../interface';
import { UserNotificationSetting } from '../models';
import {
    EmailNotificationService,
    IdentityService,
    PushNotificationService,
} from '.';

export class AppNotificationService {
    private readonly identityService: IdentityService;
    private readonly pushNotificationService: PushNotificationService;
    private readonly emailNotificationService: EmailNotificationService;

    constructor(private readonly dataSource: DataSource) {
        this.identityService = new IdentityService();
        this.pushNotificationService = new PushNotificationService();
        this.emailNotificationService = new EmailNotificationService(
            dataSource,
        );
    }

    getAllUserNotificationSettings = async (
        companyId: number,
        notificationEventType: string,
    ): Promise<UserNotificationSetting[]> => {
        return this.dataSource.getRepository(UserNotificationSetting).find({
            where: {
                company_id: companyId,
                notification_event_type: notificationEventType,
                is_deleted: false,
            },
        });
    };

    async notifyApplicationWithdraw(message: EventModel<JSON>): Promise<void> {
        const applicationWithdrawPayload: ApplicationWithdraw = <
            ApplicationWithdraw
        >(<unknown>message.data);

        const notificationEventType = message.event_type;

        const userNotificationSettings =
            await this.getAllUserNotificationSettings(
                applicationWithdrawPayload.company_id,
                notificationEventType,
            );

        if (userNotificationSettings.length === 0) {
            return;
        }

        const userAccountIds = userNotificationSettings.map(
            (i) => i.user_account_id,
        );
        const userAccountDetails =
            await this.identityService.getUsersByUserAccounts(userAccountIds);

        await this.emailNotificationService.sendEmailNotification(
            notificationEventType,
            userNotificationSettings,
            userAccountDetails,
        );
    }
}
