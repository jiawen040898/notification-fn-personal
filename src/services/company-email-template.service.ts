import { EventModel } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { EmailTemplateUsage } from '../constants';
import { CompanyCreated } from '../interface';
import { EmailTemplate } from '../models';

export class CompanyEmailTemplateService {
    constructor(private readonly dataSource: DataSource) {}

    async setupEmailTemplate(
        message: EventModel<CompanyCreated>,
    ): Promise<void> {
        const companyCreatedPayload = message.data;

        const defaultEmailTemplates = await this.dataSource
            .getRepository(EmailTemplate)
            .find({
                where: {
                    usage: EmailTemplateUsage.DEFAULT,
                    is_deleted: false,
                },
            });

        const companyEmailTemplates = defaultEmailTemplates.map((i) => {
            return <EmailTemplate>{
                ...i,
                id: undefined,
                company_id: companyCreatedPayload.company_id,
                usage: EmailTemplateUsage.CLIENT,
                created_by: message.user_account_id,
                created_at: undefined,
                updated_by: message.user_account_id,
                updated_at: undefined,
            };
        });

        await this.dataSource.transaction(async (manager) => {
            await manager.create(EmailTemplate, companyEmailTemplates);
        });
    }
}
