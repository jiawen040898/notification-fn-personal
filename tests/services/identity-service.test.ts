import { IdentityService } from '../../src/services';

jest.mock('../../src/services/identity.service', () => {
    const mBInstance = {
        getUsersByUserAccounts: jest.fn(() => [
            {
                id: 1,
                email: 'david@pulsif.me',
                first_name: 'Abu',
                last_name: 'Akao',
                userApplications: [],
            },
        ]),
    };
    const mB = jest.fn(() => mBInstance);
    return { IdentityService: mB };
});

jest.mock('@pulsifi/fn/utils/logger.util');

describe('IdentityService', () => {
    let sut: IdentityService;

    beforeAll(() => {
        sut = new IdentityService();
    });

    afterAll(() => {
        jest.resetAllMocks();
    });

    describe('getUsersByUserAccounts', () => {
        it('should return an array of user accounts given a list of ids', async () => {
            const userIds = [1];
            const result = await sut.getUsersByUserAccounts(userIds);

            expect(result).toMatchSnapshot();
        });
    });
});
