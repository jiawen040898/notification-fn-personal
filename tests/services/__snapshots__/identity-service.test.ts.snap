// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentityService getUsersByUserAccounts should return an array of user accounts given a list of ids 1`] = `
[
  {
    "email": "david@pulsif.me",
    "first_name": "Abu",
    "id": 1,
    "last_name": "Akao",
    "userApplications": [],
  },
]
`;
